packages/components/addon/components/hds/form/character-count/index.js

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { action } from '@ember/object';
+export const ID_PREFIX = 'character-count-';
+
+const NOOP = () => {};
+
+export default class HdsFormCharacterCountIndexComponent extends Component {
+  @tracked message;
+  @tracked currentLength; // The current number of characters typed into the associated input
+
+  /**
+   * @param maxLength
+   * @type {number}
+   * @default null
+   * @description The maximum number of characters allowed.
+   */
+  get maxLength() {
+    let { maxLength } = this.args;
+    return maxLength;
+  }
+
+  /**
+   * @param remainingLength
+   * @type {number}
+   * @default null
+   * @description The remaining number of characters.
+   */
+  get remainingLength() {
+    return this.maxLength - this.currentLength;
+  }
+
+  /**
+   * Determines the unique ID to assign to the element
+   * @method id
+   * @return {(string|null)} The "id" attribute to apply to the element or null, if no controlId is provided
+   */
+  get id() {
+    let { controlId } = this.args;
+    if (controlId) {
+      return `${ID_PREFIX}${controlId}`;
+    }
+    return null;
+  }
+
+  /**
+   * @param onInsert
+   * @type {function}
+   * @default () => {}
+   */
+  get onInsert() {
+    let { onInsert } = this.args;
+
+    let inputControl = document.getElementById(this.args.controlId);
+    if (inputControl && inputControl.value) {
+      this.updateMessage(inputControl);
+      inputControl.addEventListener('input', this.onInput);
+    }
```

Comment:
should we warn the consumer if they point to an element that doesn't exist? https://github.com/hashicorp/design-system/blob/main/packages/components/addon/modifiers/hds-clipboard.js#L35-L41

(counterpoint: do we expect them to use it as a standalone component? probably not, so maybe it's simply not useful in this case?)

(counter-counterpoint: they may want to use it in a generic "field" with custom input implementation?)

Reply:
lacking a valid `id` would only happen in a custom implementation where `@controlId` is provided by the consumer – and this is also the most likely scenario where a mistake can happen – so I added a similar warning, as suggested, in https://github.com/hashicorp/design-system/pull/1802/commits/b3cfdcbc184e3f97a6479ea8149ef3024acc0535#diff-369ec02c39a6239663ad5a4f7041563cc7c0c4492470465eddc134c12b0a6cd6R61-R64